### PR TITLE
PQC: ML-KEM

### LINKED ISSUE #4346: How to access ML-KEM?
Please forgive my ignorance. I thought that the current master included the updates of Kyber Round 3 to ML-KEM, but all I can find via API seems to be "Kyber-<XXX>-r3", not "ML-KEM-<XXX>".

Am I not looking in the right place, or is ML-KEM support just not there yet? If not there - could you share your plans for merging it in? Thanks!

### PR DESCRIPTION
**Pull request by reneme**
### Pull Request Dependencies

* #4373 - Mostly for the test abstractions. Will rebase again once this is merged.

### Descriptions

This builds on the existing Kyber implementation (and the flexibility added in #3387 and #4024) to provide an implementation of ML-KEM as specified in FIPS 203. The OIDs for ML-KEM are also updated to [the official definition](https://csrc.nist.gov/Projects/computer-security-objects-register/algorithm-registration#KEM).

While reviewing the implementation for consistency with the final FIPS 203 we did some minor cleanups/refactorings. Those are currently intertwined in this PR, but could be moved into an independent commit or PR.

Note that each flavor of Kyber/ML-KEM can be disabled at compile time via the modules `kyber`, `kyber_90s`, `ml_kem`.

Integration in FFI as well as into the Python bindings is also included and deprecates some of the kyber-specific FFI functions that are now replaced by algorithm-agnostic key encoding functions.

### TODOs

* [x] Currently, we use the KATs from [Kris Kwiatkowski](https://github.com/post-quantum-cryptography/KAT) which were [verified](https://groups.google.com/a/list.nist.gov/g/pqc-forum/c/s_Wez9FanHw) against the PQShield implementations.
      We should probably also look into the official ACVP tests: [#1](https://github.com/usnistgov/ACVP-Server/tree/master/gen-val/json-files/ML-KEM-encapDecap-FIPS203), [#2](https://github.com/usnistgov/ACVP-Server/tree/master/gen-val/json-files/ML-KEM-keyGen-FIPS203)
* [x] Use the private seeds rather than expanded structures for storing the ML-KEM private key (cc @FiloSottile)
  ~~We already have `private_key_bits()` and `raw_private_key_bits()`, the latter could return the seed (if it is available). We'll somehow have to distinguish serialized from non-serialized keys for loading: simply by size? 🤔~~
* [x] Look into the ACVP decaps KAT (currently disabled due to a lack of the `d||z`-type private key format)
_Either remove the tests entirely, or monitor whether an updated test vector is being released._
* [x] Go through FIPS 203 again and see if all validations are implemented (add tests)
* [x] ~~Decide on the API (see below)~~ (as with the other FIPS-PRs, let's not bloat this even more and build an API bridge as a follow up)
* [x] Squash (or at least clean up) the commit history
* [x] Remove ML-KEM specific key encoding functions from FFI/Python (once #4368 is merged)

### Open Questions re: API

Currently, there's no user-facing API that explicitly states "ML-KEM". Instead, users are expected to use the Kyber-classes with the ML-KEM parameter set. Technically, that should be fine, but I believe this will confuse people in a few years, when they didn't follow the PQC competition at the time and don't associate Kyber with ML-KEM anymore.

I see three options:

1. Do nothing, apart from guidance in documentation
2. Define aliases in a dedicated `ml_kem.h` header (e.g. by flat-out inheriting from the Kyber classes)
3. Rename all "Kyber" classes to "ML_KEM" (so that ML-KEM is the first-class API) and define aliases for Kyber (as suggested in 2)

Also, the FFI currently provides kyber-specific APIs. They should just work (tm) with ML-KEM key material as well, but like above, the names might become confusing for some users in the near future.

### Caveats

This pull request initially implemented the Initial Public Draft version of FIPS 203. It was never merged and eventually got updated to implement ML-KEM as specified in the final standard. As a result, no released version of Botan will implement ML-KEM-IPD.

Note that it is an addition! Kyber-R3 (and 90s mode) continue to work as before. 

### Example

```C++
#include <botan/pk_algs.h>
#include <botan/pubkey.h>
#include <botan/system_rng.h>

#include <cassert>
#include <iostream>

using namespace Botan;

int main() {
   // Alice
   System_RNG rng;
   auto sk = create_private_key("ML-KEM", rng, "ML-KEM-768");
   if(!sk) {
      std::cerr << "Failed to create private key\n";
      return 1;
   }
   auto pk = sk->public_key();

   // Bob
   const auto [ciphertext, shared_key_bob] = KEM_Encapsulation::destructure(PK_KEM_Encryptor(*pk, "Raw").encrypt(rng));

   // Alice
   auto shared_key_alice = PK_KEM_Decryptor(*sk, rng, "Raw").decrypt(ciphertext);

   // Test
   assert(shared_key_alice == shared_key_bob);
   std::cout << "Alice and Bob can now communicate securely!\n";
}

```

Closes #4346

### CODE REVIEW

**Comment by falko-strenzke** [comment]
@reneme, that's great news. I highly appreciate that the different versions can be controlled in the build and invoked at run-time independently. 
As to whether the ipd version should be merged into the master branch: I clearly tend say it should, for the following reasons:
- Keeping it in a separate branch leaves the work of keeping the code up to date with the master branch either to a maintainer of the user of the branch.
- The "kyber" version is already in master and released. So the addition of ML-KEM-idp does not introduce really a new problem regarding the maintenance of pre-final ML-KEM versions.
- I think the decision whether and when to deprecate them and finally drop the pre-final versions of the PQC algorithms should best be taken after the final versions have been published. Considering the case of "Keccak", where the pre-final version is still contained in Botan – due to real world implementations still using it – it seems there might be a chance that the pre-final versions of the PQC algorithms also remain in the field. If that is not the case, they can be dropped after an adequate grace period.

**Comment by reneme** [comment]
Rebased after underlying #3887 had to get some conflicts resolved.

**Comment by reneme** [comment]
With #4024 merged, this is now directly based on master and (finally) ready for prime time. The code to actually make ML-KEM-ipd happen isn't a lot, apart from the various `#ifdef`'ed blocks to weave it into the code base. So, in my opinion, this could actually be merged -- given that it is somewhat unclear when the final ML-KEM spec will come.

Right now the module is marked as 'experimental' and won't be built unless `./configure.py --enable-experimental-features`. So the Kyber situation would be as follows:

| Module | Description | Status |
| --- | --- | --- |
| `kyber` | The round 3 spec using SHA-3 (Kyber 3.1 spec) | **Stable** |
| `kyber_90s` | Same as above but using SHA-2 and AES | **Deprecated** |
| `ml_kem_ipd` | This pull request... | **Experimental** |

**Comment by randombit** [comment]
https://nvlpubs.nist.gov/nistpubs/fips/nist.fips.203.pdf

**Comment by reneme** [comment]
Judging from appendix C.2, the only relevant addition is the domain separation in the seed expansion in Algorithm 13 (KeyGen). Basically, they just append a single byte `k = {2,3,4}` (depending on the parameter set) to the `G()` hash function. That's a minimal change.

**Comment by falko-strenzke** [comment]
> Judging from appendix C.2, the only relevant addition is the domain separation in the seed expansion in Algorithm 13 (KeyGen). Basically, they just append a single byte `k = {2,3,4}` (depending on the parameter set) to the `G()` hash function. That's a minimal change.

That means in the ML-KEM ipd implementation the swapping of the matrix indices that happened in the ipd draft version were never implemented?

**Comment by reneme** [comment]
> That means in the ML-KEM ipd implementation the swapping of the matrix indices that happened in the ipd draft version were never implemented?

I had taken that as an editorial error in the IPD document. When implementing the IPD version we aligned with [the reference implementation](https://github.com/pq-crystals/kyber/commit/11d00ff1f20cfca1f72d819e5a45165c1e0a2816) and I don't think we actually swapped these indices.

**Comment by reneme** [comment]
The latest push contains a first round of adaptions (mostly updating the inline document references). As of now, I didn't come across any credible KATs to verify functional changes, albeit them being minimal.

# To do

* [ ] User-facing changes
  * [ ] rename `ml_kem_ipd` module
  * [ ] rename methods such as `mode.is_ml_kem_ipd`
  * [ ] rename `KyberMode` enum entries
  * [ ] ❓ Introduce `class ML_KEM_PublicKey` / `...PrivateKey`
         (and aliasing the existing `Kyber_...` classes for backward compatibility)
* [ ] internals
  * [ ] rename implementation classes (e.g. `ML_KEM_IPD_Encryptor`)
  * [ ] rename include guards
* [ ] Integrate updated KATs
* [ ] Incorporate the new domain separator in the seed calculation

**Comment by reneme** [comment]
This patch is now updated to implement ML-KEM as specified in FIPS 203. I updated the original pull request description accordingly and noted a few TODOs before this can eventually go in. (The pull request is open for a long time and used to implement ML-KEM-ipd.)

@randombit Your insight on the open question regarding API (Kyber vs. ML-KEM) would be appreciated. Especially about your experience with the maintenance burden of aliases in the API for backward (or in this case: forward) compatibility.

My feeling is that we would have to alias essentially all definitions in the public `kyber.h`, namely:

```C++
namespace Botan {

class KyberMode;
class Kyber_PublicKey;
class Kyber_PrivateKey;

}
```

**Comment by FiloSottile** [comment]
> * Add support for handling private keys as seeds rather than expanded structures (cc @FiloSottile)

💯 

> We already have `private_key_bits()` and `raw_private_key_bits()`, the latter could return the seed (if it is available). We'll somehow have to distinguish serialized from non-serialized keys for loading: simply by size? 🤔

Have you considered just not supporting the extended encoding? It's what I plan to do in Go. It solves the problem of how to distinguish them, and it means the seed is always available, so you don't have the problem of "what if someone loads an extended key and exports a seed".

**Comment by reneme** [comment]
> Have you considered just not supporting the extended encoding?

I certainly understand the appeal of that idea, my only concern would (obviously) be interoperability with keys that were generated and exported by another implementation that didn't get the memo. :smirk: ... the standards are out now and contain the expanded format, after all. Nevertheless, one could make an argument that moving private keys around isn't such common use case for real-world systems anyway, I guess. And adding an interface for expanded keys in a later version is always possible. :thinking:

Perhaps @randombit has an opinion on this as well?

**Comment by atreiber94** [comment]
Included the KATs from [Kris Kwiatkowski](https://github.com/post-quantum-cryptography/KAT) which were [verified](https://groups.google.com/a/list.nist.gov/g/pqc-forum/c/s_Wez9FanHw) against the PQShield implementations. They are working fine and were easy to integrate since they were already in the previous formats. I still want to also look into the ACVP tests since these are official.

**Comment by FiloSottile** [comment]
> > Have you considered just not supporting the extended encoding?
> 
> I certainly understand the appeal of that idea, my only concern would (obviously) be interoperability with keys that were generated and exported by another implementation that didn't get the memo. 😏 

FWIW, it sounds like the IETF may specify seed encodings, judging from the conversation in the LAMPS WG.

Another reason to start with seeds is that if you internally you have seeds you can always add an API to export expanded keys, but not vice versa.

**Comment by atreiber94** [comment]
ACVP tests are included now as well. I added them alongside the existing PQC KAT infrastructure since they do not seem to be generated by a provided seed.
This would also allow us to remove the "old" KATs when phasing out Kyber and solely rely on the ACVP formats for the finished standards since these are from NIST.

**Comment by FiloSottile** [comment]
An update on the private key format: the IETF reached consensus on using seeds as the private key encoding. 
https://mailarchive.ietf.org/arch/msg/spasm/KPg3lbPeNms2H5PNDBt0iJByuE0/

File: src/tests/data/pubkey/ml_kem.vec
Comment (atreiber94) [inline src/tests/data/pubkey/ml_kem.vec:1]: This change stems from the `gen_kyber_kat.py` but is not correct since these KATs are from the `post-quantum-cryptography/KAT` repository and not directly from the reference implementation.

File: doc/api_ref/pubkey.rst
Comment (FAlbertDev) [inline doc/api_ref/pubkey.rst:120]: [Arbitrary code location]
pubkey.rst:1167
```rst
Botan implements the following KEM schemes:

1. RSA
#. Kyber
#. FrodoKEM
#. McEliece
```
We can now write something like `#. ML-KEM (Kyber)`

Also pubkey.rst:1176 
```rst
Code Example: Kyber
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The code below demonstrates key encapsulation using the Kyber post-quantum scheme.

.. literalinclude:: /../src/examples/kyber.cpp
   :language: cpp
```
... can be adopted accordingly. Maybe we should change the Kyber example to an ML-KEM example and rename the file accordingly.

File: doc/api_ref/pubkey.rst
Comment (FAlbertDev) [inline doc/api_ref/pubkey.rst:144]: Arbitrary code location:

- credits.rst:166 can be updated with `ML-KEM/Kyber` instead of `Kyber`.
- FFI should not work for ML-KEM yet. But this is future work, I guess.

**Review by FAlbertDev — APPROVED** [review]
The changes look very good to me! I like the modular handling of seed-based private keys and the logic that differs between Kyber and ML-KEM. 
I've got only some minor/editorial comments :)

File: src/lib/pubkey/kyber/kyber_common/kyber.cpp
Comment (reneme) [reply src/lib/pubkey/kyber/kyber_common/kyber.cpp:167]: Yeah, I think we should make a distinction here. This is a public-facing algorithm name and it might lead to confusion in the wild.

File: src/lib/pubkey/kyber/kyber_common/kyber_keys.h
Comment (reneme) [reply src/lib/pubkey/kyber/kyber_common/kyber_keys.h:27]: It does need/produce the entire keypair though. I.e. it returns the public portion or consumes it.

File: src/lib/pubkey/kyber/kyber_common/kyber.h
Comment (reneme) [reply src/lib/pubkey/kyber/kyber_common/kyber.h:28]: Hosenträger-Gürtel-Prinzip. 😏 Doesn't hurt to have that explicitly in the code, I would say. E.g. when someone breaks or "hacks" `configure.py`.

File: src/lib/pubkey/kyber/kyber_common/kyber_symmetric_primitives.h
Comment (reneme) [reply src/lib/pubkey/kyber/kyber_common/kyber_symmetric_primitives.h:53]: Your's assumes the structure of the return value. Mine is more generic. That said: you're right. No need to be overly fancy here. I made the customization point return an optional and now it looks like this:

```C++
         if(auto domsep = seed_expansion_domain_separator(mode)) {
            return G_split<KyberSeedRho, KyberSeedSigma>(seed, *domsep);
         } else {
            return G_split<KyberSeedRho, KyberSeedSigma>(seed);
         }
```

File: src/tests/test_kyber.cpp
Comment (reneme) [reply src/tests/test_kyber.cpp:233]: Actually, we should probably register all new algorithms under `pubkey`. 😅

File: doc/api_ref/pubkey.rst
Comment (reneme) [reply doc/api_ref/pubkey.rst:143]: Good idea. Done.

File: doc/api_ref/pubkey.rst
Comment (reneme) [reply doc/api_ref/pubkey.rst:144]: Lets update the credits independently! :)

Re:FFI: The FFI is tricky, actually. `botan_privkey_view_kyber_raw_key` and `botan_pubkey_view_kyber_raw_key` should just work with ML-KEM keys as well, I think.
For `botan_privkey_load_kyber` and `botan_pubkey_load_kyber` the idea of using the buffer length as an indicator for the required parameter set is coming to bite us. The public key lengths are now ambiguous between Kyber and ML-KEM. And the private key of ML-KEM is 64 bytes, regardless of the the parameter set. Probably we should just add new methods `botan_*_mlkem_*()` and deprecate the kyber variants in due time.

File: doc/api_ref/pubkey.rst
Comment (reneme) [reply doc/api_ref/pubkey.rst:120]: Done. Also renamed the example to `ml_kem.cpp`.

File: src/lib/pubkey/kyber/ml_kem/ml_kem_impl.cpp
Comment (reneme) [reply src/lib/pubkey/kyber/ml_kem/ml_kem_impl.cpp:36]: Good catch!

File: doc/api_ref/pubkey.rst
Comment (reneme) [reply doc/api_ref/pubkey.rst]: Re:FFI, see here: https://github.com/randombit/botan/pull/3893/commits/06c817558ce72495225e80aee515c271f716e6c3

File: src/lib/ffi/ffi_pkey_algs.cpp
Comment (FAlbertDev) [inline src/lib/ffi/ffi_pkey_algs.cpp:1110]:
```suggestion
         return invoke_view_callback(view, ctx, mlkem_key->raw_public_key_bits());
```
(I know both methods do the same)

File: src/lib/pubkey/kyber/kyber_common/kyber.cpp
Comment (FAlbertDev) [inline src/lib/pubkey/kyber/kyber_common/kyber.cpp:283]: You want to check that the public key is consistent with the private key, don't you?  Isn't this only necessary for Kyber R3, since ML-KEM uses seeds?

**Comment by reneme** [comment]
@randombit This is now ready for final review. It would be amazing to have this merged before the 3.6.0 release. Please let us know if we can help in some way.

Left for future work:

* **Clean-up of the KATs**, hopefully we'll converge on some official test vector, pending availability
* **API Future-Proofing**, currently the Public API is called `Kyber_***` and we should probably provide `ML_KEM_***` aliases (see also the original PR description)

Note:

The commits that introduce `Bounded_XOF` and that alter the `PK_Keygen_Test` class are repeated in the ML-DSA PR. We'll rebase and resolve conflicts once one of them got merged.

File: src/tests/test_kyber.cpp
Comment (FAlbertDev) [inline src/tests/test_kyber.cpp:116]: Maybe resolve these ifdefs using KyberMode::is_available

**Review by FAlbertDev — COMMENTED** [review]
The changes since my last review look good to me 👍 
I left some comments

File: src/lib/pubkey/kyber/kyber_common/kyber.cpp
Comment (reneme) [reply src/lib/pubkey/kyber/kyber_common/kyber.cpp:283]: Essentially, that's what FIPS 203 requests us to do to "check the keys". But you are totally right: This is not needed when extracting the private key from the seeds.

File: src/tests/test_kyber.cpp
Comment (reneme) [reply src/tests/test_kyber.cpp:116]: Good idea! Thanks.

**Comment by reneme** [comment]
Work on it was finished on Friday, it is now awaiting review. So feature-wise it's ready to go.

Feel free to pull it into a local build if you want to start prototyping with ML-KEM. Any feedback is appreciated.

**Review by FAlbertDev — APPROVED** [review]

**Comment by tobiasbrunner** [comment]
While trying to integrate these changes into our Botan plugin I noticed (see https://github.com/strongswan/strongswan/issues/2228) that these changes allocate the seed for the key pair in a different order, first d and then z instead of the other way around, which is what FIPS-203 describes and other implementations do (and e.g. the referenced KAT vectors do as well). So it's quite awkward to test this implementation using the API. Is there a specific reason this order was chosen?

**Comment by reneme** [comment]
>  [...] these changes allocate the seed for the key pair in a different order, first d and then z instead of the other way around, which is what FIPS-203 describes and other implementations

For reference: I added my view on this [here](https://github.com/strongswan/strongswan/issues/2228#issuecomment-2399196433). TL;DR: In my opinion, downstream users should not pose assumptions on the RNG invocation pattern of any implementation of ML-KEM. For testing with KAT vectors, they should prepare a concatenation of `d || z` taken from the KAT and pass those 64 bytes into the "private key loading procedure". For Botan that would be: `Botan::Kyber_PrivateKey(d_and_z, Botan::KyberMode("ML-KEM-512"))`.

>  Is there a specific reason this order was chosen?

Not really. We followed the order of operations as stated in FIPS 203, Algorithm 19.

**Comment by tobiasbrunner** [comment]
> For testing with KAT vectors, they should prepare a concatenation of `d || z` taken from the KAT and pass those 64 bytes into the "private key loading procedure". For Botan that would be: `Botan::Kyber_PrivateKey(d_and_z, Botan::KyberMode("ML-KEM-512"))`.

OK, thanks. That sounds like a way to handle this. I've successfully used `botan_privkey_load_ml_kem()` with my original test vectors that assume a 64-byte pull from the DRBG.

> Not really. We followed the order of operations as stated in FIPS 203, Algorithm 19.

Sorry about that. I read the order in FIPS-203 wrong for some reason, it's definitely d before z (also for the inputs to Algorithm 16). So it's more the referenced KAT vectors, which allocate z before d from the DRBG, that made this awkward.

File: src/lib/ffi/ffi_pkey_algs.cpp
Comment (randombit) [inline src/lib/ffi/ffi_pkey_algs.cpp:1092]: Could we instead just define these generically for any key? Since `raw_public_key_bits` and `raw_private_key_bits` are available for a range of algorithms, and it avoids (further) algorithm specific functions. I'd rather we in fact implement the generic case and deprecate the algorithm specific ones. Discussion in #4366

Comment (reneme) [reply src/lib/ffi/ffi_pkey_algs.cpp:1092]: Yeah. Makes sense for sure. I'd propose to just remove the ML-KEM specific one from this pull request and add the generic one independently.

File: src/lib/ffi/ffi_pkey_algs.cpp
Comment (reneme) [inline src/lib/ffi/ffi_pkey_algs.cpp:1037]: Check before dereferencing!

**Comment by reneme** [comment]
Rebased after #4367 and #4372 caused conflicts.

File: src/lib/pubkey/kyber/kyber_common/kyber_keys.h
Comment (randombit) [inline src/lib/pubkey/kyber/kyber_common/kyber_keys.h:88]: We might want to consider a variant of `poison` that takes an `optional<T>` and poisons it if it has a value, as a convenience. (No need to do so in this PR)

File: src/lib/pubkey/kyber/kyber_common/kyber_symmetric_primitives.h
Comment (randombit) [inline src/lib/pubkey/kyber/kyber_common/kyber_symmetric_primitives.h:107]: Why the `array` and not just `std::optional<uint8_t>`?

Comment (randombit) [reply src/lib/pubkey/kyber/kyber_common/kyber_symmetric_primitives.h:107]: Oh maybe this just makes life easier wrt `concat`?

**Review by randombit — APPROVED** [review]

File: src/lib/pubkey/kyber/kyber_common/kyber_symmetric_primitives.h
Comment (reneme) [reply src/lib/pubkey/kyber/kyber_common/kyber_symmetric_primitives.h:107]: Yep. That's the reason. I did look into extending concat to take scalar values along with ranges but didn't find an easy enough way to allow that.

We might want to look into that later though.

File: src/lib/pubkey/kyber/kyber_common/kyber_keys.h
Comment (reneme) [reply src/lib/pubkey/kyber/kyber_common/kyber_keys.h:88]: Sounds like a good idea. I'll look at this independently.

**Pull request merged**
